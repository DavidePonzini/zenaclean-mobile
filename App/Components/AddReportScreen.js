import React, { Component } from 'react'
import { StyleSheet, View, Image, ScrollView, Alert } from 'react-native'
import { FormInput, FormLabel, FormValidationMessage, Button, Text } from 'react-native-elements'
import { widthPercentageToDP as wp, heightPercentageToDP as hp } from 'react-native-responsive-screen'
import ImagePicker from 'react-native-image-picker'
import api from '../Services/ApiService'

const options = {
  title: 'Carica foto',
  takePhotoButtonTitle: 'Scatta foto',
  chooseFromLibraryButtonTitle: 'Scegli foto da galleria',
  cancelButtonTitle: 'Annulla',
  storageOptions: {
    skipBackup: true,
    path: 'images'
  }
}

const formItems = {
  title: 'Titolo Segnalazione',
  titleErrorMessage: 'Inserisci un titolo per la segnalazione',
  descrErrorMessage: 'Inserisci una descrizione per la segnalazione',
  address: 'Indirizzo',
  descr: 'Descrizione'
}

export default class AddReportScreen extends Component {
  constructor (props) {
    super(props)

    this.state = {
      photoSource: null,
      lat: this.props.navigation.state.params.lat,
      lng: this.props.navigation.state.params.lng,
      nameFile: 'Aggiungi Foto',
      address: 'Caricamento in corso..',
      titleError: false,
      descrError: false,
      title: '',
      description: ''
    }
  }

  convertAddress = (lat, lng) => {
    return api.getAddressFromCoords({ lat, lng }, (res) => {
      this.setState({ address: res })
    })
  }

  handleTitle = (title) => {
    this.setState({ title: title })
  }

  handleDescr = (descr) => {
    this.setState({ description: descr })
  }

  componentWillMount () {
    return this.convertAddress(this.state.lat, this.state.lng)
  }

  uploadData () {
    const data = {
      description: this.state.description,
      latitude: this.state.lat,
      longitude: this.state.lng,
      timestamp: new Date().toISOString(),
      title: this.state.title
    }
    api.uploadReport(data)
      .then(() => this.showAlert())
  }

  showAlert = () => {
    let that = this
    Alert.alert(
      'Segnalazione riuscita',
      '',
      [
        { text: 'OK', onPress: () => { that.props.navigation.goBack() } }
      ],
      { cancelable: false }
    )
  }

  imageUpload () {
    ImagePicker.showImagePicker(options, (response) => {
      console.log('Response = ', response)

      if (response.didCancel) {
        console.log('User cancelled image picker')
      } else if (response.error) {
        console.log('ImagePicker Error: ', response.error)
      } else if (response.customButton) {
        console.log('User tapped custom button: ', response.customButton)
      } else {
        // const source = { uri: response.uri };

        // You can also display the image using data:
        const source = { uri: 'data:image/jpeg;base64,' + response.data }

        this.setState({
          photoSource: source,
          nameFile: response.fileName
        })

        console.log(this.state.photoSource)
      }
    })
  }

  render () {
    const { goBack } = this.props.navigation
    return (
      <ScrollView style={styles.container_scroll}>
        <View style={styles.container}>
          <FormLabel labelStyle={styles.title_label}>{formItems.title}</FormLabel>
          <FormInput inputStyle={styles.input} value={this.state.title} onChangeText={this.handleTitle} />
          { this.state.titleError && <FormValidationMessage>{formItems.titleErrorMessage}</FormValidationMessage> }

          <FormLabel labelStyle={[styles.title_label, styles.margin_top]}>{formItems.address}</FormLabel>
          <Text style={styles.place}>{this.state.address}</Text>

          <FormLabel labelStyle={[styles.title_label, styles.margin_top]}>{formItems.descr}</FormLabel>
          <FormInput inputStyle={[styles.input, styles.input_desc]} value={this.state.descr} onChangeText={this.handleDescr} multiline />
          { this.state.descrError && <FormValidationMessage>{formItems.descrErrorMessage}</FormValidationMessage> }

          <View style={[styles.view_button, styles.margin_top]}>
            <ScrollView horizontal style={{ maxWidth: wp('60%') }}>
              <FormLabel labelStyle={styles.title_label}>{this.state.nameFile}</FormLabel>
            </ScrollView>
            <Button onPress={() => this.imageUpload()} containerViewStyle={{ borderRadius: 40 }} raised buttonStyle={styles.btn_foto} title='+' />
          </View>
          <Image source={this.state.photoSource} />
          <View style={styles.view_button}>
            <Button
              raised
              containerViewStyle={{ borderRadius: 10 }}
              iconRight={{ name: 'ios-close', type: 'ionicon', color: '#FFFFFF', size: 25 }}
              buttonStyle={[styles.form_button, styles.btn_cancel]}
              onPress={() => { goBack() }}
              title='Annulla' />
            <Button
              raised
              containerViewStyle={{ borderRadius: 10 }}
              iconRight={{ name: 'md-arrow-round-up', type: 'ionicon', color: 'lightgreen', size: 20 }}
              buttonStyle={[styles.form_button, styles.btn_send]}
              disabled={!this.state.title}
              onPress={() => this.uploadData()}
              title='Invia' />
          </View>
        </View>
      </ScrollView>
    )
  }
}

const styles = StyleSheet.create({
  container_scroll: {
    flex: 1,
    width: wp('100%'),
    height: hp('100%')
  },
  container: {
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF'
  },
  title_label: {
    fontSize: 25,
    marginBottom: wp('5%'),
    maxWidth: wp('80%')
  },
  place: {
    fontSize: 20
  },
  input: {
    width: wp('80%')
  },
  input_desc: {
    maxHeight: hp('15%')
  },
  margin_top: {
    marginTop: wp('10%')
  },
  view_button: {
    flexDirection: 'row',
    marginTop: wp('10%'),
    justifyContent: 'center',
    alignItems: 'center'
  },
  form_button: {
    width: wp('35%'),
    borderRadius: 10,
    justifyContent: 'center',
    alignItems: 'center'
  },
  btn_cancel: {
    backgroundColor: '#D32F2F'
  },
  btn_send: {
    backgroundColor: 'green'
  },
  btn_foto: {
    width: wp('15%'),
    height: wp('15%'),
    borderRadius: 40,
    backgroundColor: '#BDBDBD'
  }

})
