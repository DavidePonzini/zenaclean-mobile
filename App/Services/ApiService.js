import Secrets from 'react-native-config'

const baseUrl = Secrets.API_URL

const getAddressFromCoords = ({ lat, lng }, cb) => {
  const googleApiUrl = 'https://maps.googleapis.com/maps/api/geocode/json?address='

  fetch(googleApiUrl + lat + ',' + lng + '&key=' + Secrets.GOOGLE_MAPS_API_KEY)
    .then((response) => response.json())
    .then((responseJson) => cb(responseJson))
}

const getMarkers = (cb) => {
  fetch(baseUrl + 'markers.json')
    .then((response) => response.json())
    .then((responseJson) => { cb(responseJson) })
}

const uploadReport = (infoReport) => {
  return fetch(baseUrl + 'markers.json', {
    method: 'POST',
    body: JSON.stringify(infoReport)
  })
}
export default {
  getAddressFromCoords,
  getMarkers,
  uploadReport
}
