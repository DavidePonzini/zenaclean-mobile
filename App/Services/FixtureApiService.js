const SSNlist = ['LNGSVV60T17H509N', 'HTHHLG90R62B112N', 'BRTHZG56P68C982U']
const emailList = ['prova@gmail.com', 'prova@hotmail.it']
export const user1 = {
  email: emailList[0],
  password: '12345678',
  ssn: SSNlist[0],
  id: '5c5afd7c4be7ff1c8d7daacd'
}

const wallets = {}
wallets[user1.id] = 2
export { wallets }
export const fixtureAddress = 'Via del Camoscio'
export const fixtureMarkers = [{ address: fixtureAddress, title: 'Divano abbandonato', latitude: 44.1234, longitude: 8.1234, description: 'Divano abbandonato in via Gramsci', timestamp: '2018-12-09T14:40:35 -02:00', approved_positive: false, approved_negative: false, _id: '5c6580a33a7f647afccd8920', user_id: user1.id, url: 'images/667856780e39f0bbe7407bfdae603f8e', voted_positive: false, voted_negative: false }]
export const fixtureMessageOk = 'Registrazione effettuata con successo!'
export const fixtureMessageNotOk = 'Utente giÃ  esistente'
export const fixtureUserId = 'verylonguserid'
const totVoti = [32, 45, 21]
export let userId = null

const logInUser = (email, password, cb) => {
  return new Promise(resolve => {
    (user1.email === email && user1.password === password) ? resolve({ status: 'ok' }) : resolve({ status: 'no' })
  }).then((res) => {
    if (res.status === 'ok') {
      userId = fixtureUserId
      cb(null, { id: userId, email, eth_address: null })
    } else {
      cb(new Error('login failed'), res)
    }
  })
}

const getAddressFromCoords = ({ lat, lng }, cb) => {
  return new Promise((resolve) => {
    resolve(fixtureAddress)
  }).then((res) => { cb(null, res) })
}

const getMarkers = (neLat, swLat, swLng, neLng, cb) => {
  return new Promise((resolve) => {
    resolve(JSON.stringify(fixtureMarkers))
  }).then((res) => { cb(JSON.parse(res)) })
}

const uploadReport = (infoReport, cb) => {
  return new Promise((resolve) => {
    resolve({ status: 'ok' })
  }).then(() => {
    cb(null)
  })
}

const registerUser = (email, ssn, password, cb) => {
  return new Promise((resolve) => {
    let checked = false
    for (let i = 0; i < SSNlist.length; i++) {
      if (SSNlist[i] === ssn) checked = true
    }
    for (let i = 0; i < emailList.length; i++) {
      if (emailList[i] === email) checked = true
    }
    if (!checked) { resolve(fixtureMessageOk) }
    if (checked) { resolve(fixtureMessageNotOk) }
  }).then((res) => { cb(null, res) })
}

const voteReport = (v, report, cb) => {
  return new Promise((resolve) => {
    let old = totVoti[report]
    let current = totVoti[report] + 1
    if (old < current) resolve({ status: 'ok' })
    else resolve({ status: 'no' })
  }).then((res, current) => {
    if (res.status === 'ok') {
      totVoti[report] = current
      cb(null)
    } else {
      cb(new Error('login failed'), res)
    }
  })
}

const changePassword = (email, oldPassword, newPassword, confirmPassword, cb) => {
  return new Promise((resolve) => {
    if (user1.password === oldPassword && newPassword === confirmPassword) resolve({ status: 'ok' })
    else resolve({ status: 'no' })
  }).then((res) => {
    if (res.status === 'ok') {
      user1.password = newPassword
      cb(null)
    } else {
      cb(new Error('login failed'), res)
    }
  })
}

const isLoggedIn = () => userId != null

const logoutUser = () => {
  userId = null
}
const rehydrateLogin = () => {
  return new Promise((resolve, reject) => {
    if (userId != null) {
      resolve(userId)
    } else {
      reject(new Error('Not logged in'))
    }
  })
}

const getBalance = (eth_addr, cb) => {
  return new Promise(resolve => {
    resolve({ value: wallets[user1.id] })
  }).then(res => {
    cb(res.value)
  })
}

export default {
  getAddressFromCoords,
  getMarkers,
  uploadReport,
  registerUser,
  logInUser,
  changePassword,
  voteReport,
  isLoggedIn,
  logoutUser,
  rehydrateLogin,
  getBalance
}
