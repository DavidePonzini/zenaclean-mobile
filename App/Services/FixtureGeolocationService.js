const fixturePosition = {
  coords: {
    accuracy: 1135,
    altitude: null,
    altitudeAccuracy: null,
    heading: null,
    latitude: 44.4135423,
    longitude: 8.9554223,
    speed: null
  },
  timestamp: 1547641567823
}

const fixtureRegion = {
  latitude: 44.4135423,
  longitude: 8.9554223,
  latitudeDelta: 0.1,
  longitudeDelta: 0.1
}

let lastLocation = fixturePosition
let currentRegion = fixtureRegion

const setCurrentRegion = reg => { currentRegion = reg }
const getCurrentRegion = () => currentRegion

const geolocateOnce = async () => {
  return new Promise(resolve => {
    let pos = fixturePosition
    lastLocation = pos
    resolve(pos)
  })
}

const getLatestPosition = () => lastLocation

const geolocateContinuously = (cb, cbErr, options) => {
  return setInterval(cb, 100)
}

const stopGeolocating = (id) => {
  clearInterval(id)
}

export default {
  geolocateOnce,
  geolocateContinuously,
  stopGeolocating,
  getLatestPosition,
  getCurrentRegion,
  setCurrentRegion
}
