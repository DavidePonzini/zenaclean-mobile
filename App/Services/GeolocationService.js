let lastLocation = null

// initial region
let currentRegion = {
  latitude: 44.4056,
  longitude: 8.9463,
  latitudeDelta: 0.1,
  longitudeDelta: 0.1
}

const setCurrentRegion = reg => { currentRegion = reg }
const getCurrentRegion = () => currentRegion

const geolocateOnce = async () => {
  return new Promise(resolve => {
    navigator.geolocation.getCurrentPosition((pos) => {
      lastLocation = pos
      resolve(pos)
    })
  })
}

const geolocateContinuously = (cb, cbErr, options) => {
  return navigator.geolocation.watchPosition((pos) => {
    cb(pos)
    lastLocation = pos
  }, cbErr, options)
}

const stopGeolocating = (id) => {
  navigator.geolocation.clearWatch(id)
}

const getLatestPosition = () => lastLocation

export default {
  geolocateOnce,
  geolocateContinuously,
  stopGeolocating,
  getLatestPosition,
  getCurrentRegion,
  setCurrentRegion
}
