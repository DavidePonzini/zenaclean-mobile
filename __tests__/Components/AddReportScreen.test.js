import React from 'react'
import AddReportScreen from '../../App/Components/AddReportScreen'

import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import 'isomorphic-fetch'
import { shallowToJson } from 'enzyme-to-json'
import { fixtureMarkers } from '../../App/Services/FixtureApiService'

configure({ adapter: new Adapter() })

const fixtureNavigation = { navigate: jest.fn(),
  state: {
    params: {
      lat: fixtureMarkers[0].latitude,
      lng: fixtureMarkers[0].longitude
    }
  }
}

jest.mock('../../App/Services/ApiService', () => require('../../App/Services/FixtureApiService'))

describe('AddReportScreen tests', () => {
  const wrapper = shallow(<AddReportScreen navigation={fixtureNavigation} />)
  it('renders correctly', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })
  const instance = wrapper.instance()
  it('converts coordinates to address', async () => {
    await instance.componentWillMount()
    expect(instance.state.address).toBe('Via del Camoscio')
  })
  const submitButton = wrapper.findWhere((element) => element.props().title === 'Invia').first()
  it('initializes title with empty string', () => {
    const titleComponent = wrapper.findWhere((element) => element.props().onChangeText === instance.handleTitle).first()
    expect(titleComponent.props().value).toEqual('')
  })
  it('prevents submit with empty title', () => {
    expect(submitButton.props().disabled).toBeTruthy()
  })
  it('uploads data successfully', async () => {
    instance.uploadData = jest.fn()
    instance.handleTitle('Divano abbandonato')
    wrapper.update()
    submitButton.simulate('Press')
    expect(instance.uploadData).toHaveBeenCalled()
  })
})
