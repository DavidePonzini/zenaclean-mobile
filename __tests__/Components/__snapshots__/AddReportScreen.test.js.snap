// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddReportScreen tests renders correctly 1`] = `
<ScrollView
  style={
    Object {
      "flex": 1,
      "height": 1334,
      "width": 750,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5FCFF",
        "justifyContent": "center",
      }
    }
  >
    <FormLabel
      labelStyle={
        Object {
          "fontSize": 25,
          "marginBottom": 37.5,
          "maxWidth": 600,
        }
      }
    >
      Titolo Segnalazione
    </FormLabel>
    <FormInput
      inputStyle={
        Object {
          "width": 600,
        }
      }
      normalizeFontSize={true}
      onChangeText={[Function]}
      value=""
    />
    <FormLabel
      labelStyle={
        Array [
          Object {
            "fontSize": 25,
            "marginBottom": 37.5,
            "maxWidth": 600,
          },
          Object {
            "marginTop": 75,
          },
        ]
      }
    >
      Indirizzo
    </FormLabel>
    <TextElement
      style={
        Object {
          "fontSize": 20,
        }
      }
    >
      Caricamento in corso..
    </TextElement>
    <FormLabel
      labelStyle={
        Array [
          Object {
            "fontSize": 25,
            "marginBottom": 37.5,
            "maxWidth": 600,
          },
          Object {
            "marginTop": 75,
          },
        ]
      }
    >
      Descrizione
    </FormLabel>
    <FormInput
      inputStyle={
        Array [
          Object {
            "width": 600,
          },
          Object {
            "maxHeight": 200,
          },
        ]
      }
      multiline={true}
      normalizeFontSize={true}
      onChangeText={[Function]}
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
            "marginTop": 75,
          },
          Object {
            "marginTop": 75,
          },
        ]
      }
    >
      <ScrollView
        horizontal={true}
        style={
          Object {
            "maxWidth": 450,
          }
        }
      >
        <FormLabel
          labelStyle={
            Object {
              "fontSize": 25,
              "marginBottom": 37.5,
              "maxWidth": 600,
            }
          }
        >
          Aggiungi Foto
        </FormLabel>
      </ScrollView>
      <Button
        buttonStyle={
          Object {
            "backgroundColor": "#BDBDBD",
            "borderRadius": 40,
            "height": 112.5,
            "width": 112.5,
          }
        }
        containerViewStyle={
          Object {
            "borderRadius": 40,
          }
        }
        onPress={[Function]}
        raised={true}
        title="+"
      />
    </View>
    <Component
      source={null}
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "marginTop": 75,
        }
      }
    >
      <Button
        buttonStyle={
          Array [
            Object {
              "alignItems": "center",
              "borderRadius": 10,
              "justifyContent": "center",
              "width": 262.5,
            },
            Object {
              "backgroundColor": "#D32F2F",
            },
          ]
        }
        containerViewStyle={
          Object {
            "borderRadius": 10,
          }
        }
        iconRight={
          Object {
            "color": "#FFFFFF",
            "name": "ios-close",
            "size": 25,
            "type": "ionicon",
          }
        }
        onPress={[Function]}
        raised={true}
        title="Annulla"
      />
      <Button
        buttonStyle={
          Array [
            Object {
              "alignItems": "center",
              "borderRadius": 10,
              "justifyContent": "center",
              "width": 262.5,
            },
            Object {
              "backgroundColor": "green",
            },
          ]
        }
        containerViewStyle={
          Object {
            "borderRadius": 10,
          }
        }
        disabled={true}
        iconRight={
          Object {
            "color": "lightgreen",
            "name": "md-arrow-round-up",
            "size": 20,
            "type": "ionicon",
          }
        }
        onPress={[Function]}
        raised={true}
        title="Invia"
      />
    </View>
  </View>
</ScrollView>
`;
